test/addons-napi/test_promise/test_promise.c

```diff
@@ -0,0 +1,50 @@
+#include <stdio.h>
+#include <node_api.h>
+#include "../common.h"
+
+napi_value bind_napi_create_promise(napi_env env, napi_callback_info info) {
+  napi_value promise, resolve, reject, result;
+
+  NAPI_CALL(env, napi_create_promise(env, &promise, &resolve, &reject));
+  NAPI_CALL(env, napi_create_object(env, &result));
+
+  fprintf(stderr, "resolve: %p\n", resolve);
+
+  napi_valuetype value_type;
+  NAPI_CALL(env, napi_typeof(env, resolve, &value_type));
+  fprintf(stderr, "resolve: value_type: %d\n", value_type);
+
+  napi_property_descriptor properties[] = {
+    { "promise", NULL, NULL, NULL, NULL, promise, napi_default, NULL },
+    { "resolve", NULL, NULL, NULL, NULL, resolve, napi_default, NULL },
+    { "reject", NULL, NULL, NULL, NULL, reject, napi_default, NULL },
+  };
```

Comment:
... and in V8 at least, the deferred has engine support for implementing `napi_resolve_deferred()` and `napi_reject_deferred()` in the form of `v8::Promise::Resolver::Resolve()` resp. `v8::Promise::Resolver::Reject()`.

In ChakraCore I suspect the deferred will take the form of a JS object with two methods.